src/views/primary-auth/PrimaryAuthForm.js

```diff
@@ -38,26 +38,37 @@ define([
     initialize: function () {
       this.listenTo(this, 'save', function () {
         var processCreds = this.settings.get('processCreds');
-        if (_.isFunction(processCreds)) {
-          processCreds({
-            username: this.model.get('username'),
-            password: this.model.get('password')
-          });
+        var asyncProcessCreds = this.settings.get('asyncProcessCreds');
+        var self = this;
+        if(_.isFunction(asyncProcessCreds)) {
+            asyncProcessCreds({
+              username: this.model.get('username'),
+              password: this.model.get('password')
+            }, function(){ postProcessCreds(); });
+        }else{
+          if (_.isFunction(processCreds)) {
+            processCreds({
+              username: this.model.get('username'),
+              password: this.model.get('password')
+            });
+          }
+          postProcessCreds();
         }
-        if (this.settings.get('features.deviceFingerprinting')) {
-          var self = this;
-          DeviceFingerprint.generateDeviceFingerprint(this.settings.get('baseUrl'), this.$el)
-          .then(function (fingerprint) {
-            self.model.set('deviceFingerprint', fingerprint);
-          })
-          .fail(function () {
-            // Keep going even if device fingerprint fails
-          })
-          .fin(function () {
+        function postProcessCreds() {
+          if (this.settings.get('features.deviceFingerprinting')) {
```

Comment:
I will also refactor to promises. I might be a while getting back with the updated/tested version. I'll make sure to test DeviceFingerprint as well to ensure I haven't broken that on the way.